### PR TITLE
Keras-like API compute outputshape, core layers, conv+pooling layers

### PR DESCRIPTION
**Pull request by hkvision**
## What changes were proposed in this pull request?

ComputeOutputShape logic for almost all layers.
Added some core layers, convolution layers, pooling layers compatible with Keras.

Layers:
Convolution1D
Convolution3D
MaxPooling1D
MaxPooling3D
AveragePooling1D
AveragePooling2D
AveragePooling3D
GlobalMaxPooling2D
GlobalAveragePooling2D
RepeatVector
Permute
Cropping1D
Cropping2D
Cropping3D

## How was this patch tested?

Jenkins

### CODE REVIEW

File: spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/keras/Convolution1D.scala [hunk 1]
```diff
@@ -0,0 +1,112 @@
...
+    bRegularizer: Regularizer[T] = null,
+    bias: Boolean = true,
+    inputShape: Shape = null)(implicit ev: TensorNumeric[T]) : Convolution1D[T] = {
+    new Convolution1D[T](nbFilter, filterLength,
+      KerasUtils.getInitMethod(init), KerasUtils.getActivation(activation),
+      borderMode, subsampleLength, wRegularizer, bRegularizer, bias, inputShape)
+  }
+}
+
+object Conv1D {
```
Comment (hkvision) [inline spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/keras/Convolution1D.scala]: Need to move alias to package object

**Comment by hkvision** [comment]
To be discussed: for stride=None, we use null or Option? @zhichao-li

**Comment by hkvision** [comment]
@zhichao-li pls have a review.

**Comment by hkvision** [comment]
jenkins passed http://172.168.2.101:8080/job/BigDL-PR-Validation-v3/52/

File: spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/keras/Pooling1D.scala [hunk 2]
```diff
@@ -0,0 +1,42 @@
...
+
+import com.intel.analytics.bigdl.tensor.Tensor
+import com.intel.analytics.bigdl.tensor.TensorNumericMath.TensorNumeric
+import com.intel.analytics.bigdl.utils.Shape
+
+import scala.reflect.ClassTag
+
+abstract class Pooling1D[T: ClassTag](
+   val poolLength: Int = 2,
+   var stride: Option[Int] = None,
```
Comment (hkvision) [inline spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/keras/Pooling1D.scala]: val

**Comment by zhichao-li** [comment]
enrich all of the error message with the essential variable. i.e input length

**Comment by zhichao-li** [comment]
For `stride`, let's just simply use Int here and replace None with its default value. i.e poolLength.

File: spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/keras/Pooling1D.scala (hunk 2 above)
Comment (zhichao-li) [reply spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/keras/Pooling1D.scala]: let's just simply use Int here and replace None with its default value. i.e poolLength.

File: spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/SpatialFullConvolution.scala [hunk 3]
```diff
@@ -252,6 +252,15 @@ class SpatialFullConvolution[T: ClassTag](
     }
   }
 
+  override def computeOutputShape(inputShape: Shape): Shape = {
```
Comment (zhichao-li) [inline spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/SpatialFullConvolution.scala]: Do you use this in keras layer? let's remove it if you have override `computeOutputShape` in keras. The idea is we only touch the code if only necessary for the keras layer.

File: spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/keras/Pooling1D.scala [hunk 4]
```diff
@@ -0,0 +1,44 @@
...
+import scala.reflect.ClassTag
+
+abstract class Pooling1D[T: ClassTag](
+   val poolLength: Int = 2,
+   var stride: Int = -1,
+   val borderMode: String = "valid",
+   var inputShape: Shape = null)(implicit ev: TensorNumeric[T])
+  extends KerasLayer[Tensor[T], Tensor[T], T](KerasLayer.addBatch(inputShape)) {
+
+  require(stride == -1 || stride > 0, s"Invalid stride value for Pooling1D: $stride")
```
Comment (zhichao-li) [inline spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/keras/Pooling1D.scala]: ??

File: spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/keras/Pooling1D.scala [hunk 5]
```diff
@@ -0,0 +1,44 @@
...
+  def strideValue: Int = if (stride > 0) stride else poolLength
```
Comment (zhichao-li) [inline spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/keras/Pooling1D.scala]: can it be a val here?

File: spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/keras/Pooling1D.scala [hunk 6]
```diff
@@ -0,0 +1,44 @@
...
+
+import com.intel.analytics.bigdl.tensor.Tensor
+import com.intel.analytics.bigdl.tensor.TensorNumericMath.TensorNumeric
+import com.intel.analytics.bigdl.utils.Shape
+
+import scala.reflect.ClassTag
+
+abstract class Pooling1D[T: ClassTag](
+   val poolLength: Int = 2,
+   var stride: Int = -1,
```
Comment (zhichao-li) [inline spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/keras/Pooling1D.scala]: should be val here?

File: spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/keras/Pooling1D.scala [hunk 7]
```diff
@@ -0,0 +1,44 @@
...
+   val borderMode: String = "valid",
+   var inputShape: Shape = null)(implicit ev: TensorNumeric[T])
```
Comment (zhichao-li) [inline spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/keras/Pooling1D.scala]: val?

**Comment by hkvision** [comment]
All the comments are met. @zhichao-li

**Comment by zhichao-li** [comment]
LGTM

**Comment by hkvision** [comment]
jenkins passed http://172.168.2.101:8080/view/PR-V(new)/job/BigDL-PR-Validation-v3/135/

**Pull request merged**